### DISCUSSION TITLE
How to get subscriber info via OPCUA Server

### DISCUSSION THREAD

**Discussion by sangjiexun** [eclipse-milo/milo#957] [General]
Dear developer, how can I get the subscriber (OPCUA Client) information from OPCUA Server, such as remote device's ip ,address and  computer-name, etc.

**Comment by kevinherron** [comment]
If you're using the server SDK then the `Session` object has a `getClientAddress` method on it: https://github.com/eclipse/milo/blob/f5799292024a9fb8c9a025e0664c49d67c1110bf/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/Session.java#L251-L260

**Comment by sangjiexun** [comment]
> If you're using the server SDK then the `Session` object has a `getClientAddress` method on it:
> 
> https://github.com/eclipse/milo/blob/f5799292024a9fb8c9a025e0664c49d67c1110bf/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/Session.java#L251-L260

THANKS!!!

**Comment by parvit** [comment]
@kevinherron Hi, using the `getClientAddress` method during a `onSessionCreated` callback returns a null value. How is it possible to detect on the server when a client has activated the connection and get the correct remote address? Thanks

**Reply by kevinherron** [reply]
It isn't set until the session is activated.

I think SessionListener in the server was meant to be used internally, by the diagnostics implementation. I'm not sure what use it would be to you as an implementor.

**Reply by parvit** [reply]
The use case for this is to track the connection and disconnection of clients, the behavior of the library here is strange because when the `onSessionClosed` callback is invoked, the session object does have the address set.

We work around with a scheduled runner on the server to check the list of sessions, but it would have been nicer to have the callbacks report the correct data. Thanks anyway for the support.

**Reply by kevinherron** [reply]
It's not that strange, a Session that has been closed was probably activated at some point.

Certain information is set or only available when the Session is activated: https://github.com/eclipse/milo/blob/3419235853491ac89f3391bffd337c8200522522/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/SessionManager.java#L539-L542

**Reply by kevinherron** [reply]
That said, I think that the address _could_ be set earlier, however whatever you're trying to do tracking Sessions/Clients might be faulty, because you are probably only interested in  _activated_ Sessions, and there is no callback for that.
